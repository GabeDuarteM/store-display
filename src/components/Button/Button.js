import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

export const styles = {
  button: {
    cursor: ({ disabled }) => (disabled ? 'not-allowed' : 'pointer'),
    width: ({ width }) => width,
    height: 50,
    backgroundColor: '#6e8be3',
    userSelect: 'none',
    color: '#ffffff',
    fontSize: 14,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 3,
    border: 0,
  },
}

const Button = ({
  classes,
  className,
  children,
  onClick,
  disabled,
  ...rest
}) => (
  <button
    className={cn(classes.button, className)}
    onClick={onClick}
    disabled={disabled}
    {...rest}
  >
    {children}
  </button>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** The content of the button */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf([PropTypes.node]),
    PropTypes.node,
  ]).isRequired,
  /** The click handler of the button. */
  onClick: PropTypes.func.isRequired,
  /** If present, it will be added to the root component */
  className: PropTypes.string,
  /** The width of the component */
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]), // eslint-disable-line react/no-unused-prop-types
  /** Prop that defines if the button is disabled or not */
  disabled: PropTypes.bool,
}

export const defaultProps = {
  width: 280,
  disabled: false,
  className: null,
}

Button.propTypes = propTypes
Button.defaultProps = defaultProps

export default injectSheet(styles)(Button)
