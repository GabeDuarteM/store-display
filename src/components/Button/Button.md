```js
<Button onClick={() => {}}>Load more</Button>
```

```js
const loading = require('../../assets/loading.svg')
;<div>
  <Button disabled onClick={() => {}}>
    <SvgIcon width={35} src={loading} />
  </Button>
</div>
```
