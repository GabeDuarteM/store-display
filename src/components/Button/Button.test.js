import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import Button, { styles, propTypes } from './Button'

const createWrapper = (func, { children, ...optionalProps }) =>
  func(
    <Button onClick={() => {}} {...optionalProps}>
      {children || 'Button text'}
    </Button>,
  )

const createWrapperShallow = (optionalProps = {}) =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = (optionalProps = {}) =>
  createWrapper(mount, optionalProps)

describe('COMPONENT: <Button />', () => {
  it('should render without crashing with default props', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's style snapshot", () => {
    expect(styles).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it('should apply whatever classname that is passed as props', () => {
    const className = 'custom-classname'
    const wrapper = createWrapperShallow({ className })

    const hasSpecifiedClassname = wrapper.dive().hasClass(className)
    expect(hasSpecifiedClassname).toBe(true)
  })

  it('should call onClick when the user clicks the button', () => {
    const onClick = jest.fn()
    const wrapper = createWrapperShallow({ onClick })

    wrapper.dive().simulate('click')
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('should pass the disabled state to the button', () => {
    const disabled = true
    const wrapper = createWrapperShallow({ disabled })

    const disabledProp = wrapper.dive().prop('disabled')
    expect(disabledProp).toBe(true)
  })

  it('should render whatever children that is passed', () => {
    const children = 'Some random button'
    const wrapper = createWrapperShallow({ children })

    const childrenText = wrapper.dive().text()
    expect(childrenText).toBe(children)
  })
})
