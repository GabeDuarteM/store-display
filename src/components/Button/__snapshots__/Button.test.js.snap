// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <Button /> should match component's propTypes snapshot 1`] = `
Object {
  "children": [Function],
  "className": [Function],
  "classes": [Function],
  "disabled": [Function],
  "onClick": [Function],
  "width": [Function],
}
`;

exports[`COMPONENT: <Button /> should match component's snapshot 1`] = `
<button
  className="Button-button Button-button"
  disabled={false}
  onClick={[Function]}
  width={280}
>
  Button text
</button>
`;

exports[`COMPONENT: <Button /> should match component's style snapshot 1`] = `
Object {
  "button": Object {
    "alignItems": "center",
    "backgroundColor": "#6e8be3",
    "border": 0,
    "borderRadius": 3,
    "color": "#ffffff",
    "cursor": [Function],
    "display": "flex",
    "fontSize": 14,
    "height": 50,
    "justifyContent": "center",
    "userSelect": "none",
    "width": [Function],
  },
}
`;
